components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/RoleBasedScopeIssuer.java

```diff
@@ -0,0 +1,758 @@
+            // If GrantType is SAML20_BEARER and CHECK_ROLES_FROM_SAML_ASSERTION is true, or if GrantType is
+            // JWT_BEARER and retrieveRolesFromUserStoreForScopeValidation system property is true,
+            // use user roles from assertion or jwt otherwise use roles from user store.
+            String isSAML2Enabled = System.getProperty(OAuth2Constants.RoleBasedScope.CHECK_ROLES_FROM_SAML_ASSERTION);
+            String isRetrieveRolesFromUserStoreForScopeValidation = System
+                    .getProperty(OAuth2Constants.RoleBasedScope.RETRIEVE_ROLES_FROM_USERSTORE_FOR_SCOPE_VALIDATION);
+            if (GrantType.SAML20_BEARER.toString().equals(grantType) && Boolean.parseBoolean(isSAML2Enabled)) {
+                authenticatedUser.setUserStoreDomain("FEDERATED");
+                tokReqMsgCtx.setAuthorizedUser(authenticatedUser);
+                Assertion assertion = (Assertion) tokReqMsgCtx
+                        .getProperty(OAuth2Constants.RoleBasedScope.SAML2_ASSERTION);
+                userRoles = getRolesFromAssertion(assertion);
+            } else if (OAuthConstants.GrantTypes.JWT_BEARER.equals(grantType) && !(Boolean
+                    .parseBoolean(isRetrieveRolesFromUserStoreForScopeValidation))) {
+                configureForJWTGrant(tokReqMsgCtx);
+                Map<ClaimMapping, String> userAttributes = authenticatedUser.getUserAttributes();
+                if (tokReqMsgCtx.getProperty(OAuth2Constants.RoleBasedScope.ROLE_CLAIM) != null) {
+                    userRoles = getRolesFromUserAttribute(userAttributes,
+                            tokReqMsgCtx.getProperty(OAuth2Constants.RoleBasedScope.ROLE_CLAIM).toString());
+                }
+            } else {
+                if (authenticatedUser.isFederatedUser() && StringUtils.equals(REFRESH_TOKEN_GRANT_TYPE, grantType)) {
+                    try {
+                        authenticatedUser.setUserName(OAuth2Util.getAppInformationByClientId(clientId).getAppOwner().
+                                getUserName());
+                    } catch (InvalidOAuthClientException | IdentityOAuth2Exception e) {
+                        log.error("Error when retrieving the username " + e.getMessage(), e);
+                    }
+                }
+                userRoles = getUserRoles(authenticatedUser, grantType);
+            }
+            authorizedScopes = getAuthorizedScopes(userRoles, requestedScopes, appScopes);
+            scopes.addAll(authorizedScopes);
+        }
+        return scopes;
+    }
+
+    /**
+     * This method is used to check whether a given scope is a product rest API scope or not.
+     *
+     * @param scope scope
+     * @return true if it is a product rest API scope
+     */
+    private boolean checkForProductRestAPIScopes(String scope) {
+
+        return scope.startsWith("apim:") || scope.startsWith("apim_analytics:") ||
+                scope.startsWith("service_catalog:");
```

Comment:
let's use constants for these. or better if we can have a list for the prefixes

something similar to

```
private static final List<String> ALLOWED_PREFIXES = Arrays.asList("apim:", "apim_analytics:", "service_catalog:");

public boolean isScopeAllowed(String scope) {
    for (String prefix : ALLOWED_PREFIXES) {
        if (scope.startsWith(prefix)) {
            return true;
        }
    }
    return false;
}
```

Reply:
fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2084/commits/ad343f6bb0ec0f4978d6e599ac94b1835dab76de using constants.   introducing a loop will add unnecessary overhead.